File: YukariLauncher/src/main/java/net/kdt/pojavlaunch/customcontrols/LayoutSanitizer.java
package net.kdt.pojavlaunch.customcontrols;

import java.util.Iterator;
import java.util.List;

public class LayoutSanitizer {
/**
 * このオブジェクトが「InvalidFormula」状態であるかを判定します。
 */

    // Maybe add more conditions here later?
    private static boolean isInvalidFormula(String formula) {
        return formula.contains("Infinity");
    }
/**
 * このオブジェクトが「SaneData」状態であるかを判定します。
 */
    private static boolean isSaneData(ControlData controlData) {
        if(controlData.getWidth() == 0 || controlData.getHeight() == 0) return false;
        if(isInvalidFormula(controlData.dynamicX) || isInvalidFormula(controlData.dynamicY)) return false;
        return true;
    }
/**
 * 「ControlData」の値を取得します。
 */
    private static ControlData getControlData(Object dataEntry) {
        if(dataEntry instanceof ControlData) {
            return (ControlData) dataEntry;
        }else if(dataEntry instanceof ControlDrawerData) {
            return ((ControlDrawerData) dataEntry).properties;
        }else throw new RuntimeException("Encountered wrong type during ControlData sanitization");
    }
/**
 * 「sanitize List」処理を実行します。
 * このメソッドは特定の機能を提供するために実装されています。
 */
    private static boolean sanitizeList(List<?> controlDataList) {
        boolean madeChanges = false;
        Iterator<?> iterator = controlDataList.iterator();
        while(iterator.hasNext()) {
            ControlData controlData = getControlData(iterator.next());
            if(!isSaneData(controlData)) {
                madeChanges = true;
                iterator.remove();
            }
        }
        return madeChanges;
    }

    /**
     * コントロールレイアウト内のすべてのボタンをチェックし、適切な値を持つことを確認します
     * to be displayed properly). Removes any buttons deemed not sane.
     * @param controls the original control layout.
     * @return whether the sanitization process made any changes to the layout
     */
    public static boolean sanitizeLayout(CustomControls controls) {
        boolean madeChanges = sanitizeList(controls.mControlDataList);
        if(sanitizeList(controls.mDrawerDataList)) madeChanges = true;
        if(sanitizeList(controls.mJoystickDataList)) madeChanges = true;
        return madeChanges;
    }
}

File: YukariLauncher/src/main/java/com/arata/yukarilauncher/ui/subassembly/customcontrols/LayoutSanitizer.kt
package com.arata.yukarilauncher.ui.subassembly.customcontrols

/** コントロールレイアウトデータを検証し、無効なエントリを除去するユーティリティオブジェクト。 */
object LayoutSanitizer {
    /**
     * 数式にInfinityが含まれているかチェックします。
     * @param formula チェックする数式文字列
     * @return Infinityを含む場合はtrue
     */
    private fun isInvalidFormula(formula: String?): Boolean {
        return formula?.contains("Infinity") == true
    }

    /**
     * ControlDataが有効な値を持っているか判定します。
     * @param controlData 検証するデータ
     * @return 有効な場合はtrue
     */
    private fun isSaneData(controlData: ControlData): Boolean {
        if (controlData.getWidth() == 0f || controlData.getHeight() == 0f) return false
        if (isInvalidFormula(controlData.dynamicX) || isInvalidFormula(controlData.dynamicY)) return false
        return true
    }

    /**
     * データエントリからControlDataを取得します。
     * @param dataEntry ControlDataまたはControlDrawerData
     * @return 抽出されたControlData
     */
    private fun getControlData(dataEntry: Any): ControlData {
        return when (dataEntry) {
            is ControlData -> dataEntry
            is ControlDrawerData -> dataEntry.properties
            else -> throw RuntimeException("Encountered wrong type during ControlData sanitization")
        }
    }

    /**
     * リスト内の無効なデータを除去します。
     * @param controlDataList 検証するデータのリスト
     * @return 変更があった場合はtrue
     */
    private fun sanitizeList(controlDataList: MutableList<*>): Boolean {
        var madeChanges = false
        val iterator = controlDataList.iterator()
        while (iterator.hasNext()) {
            val controlData = getControlData(iterator.next()!!)
            if (!isSaneData(controlData)) {
                madeChanges = true
                iterator.remove()
            }
        }
        return madeChanges
    }

    /**
     * CustomControls全体を検証し、無効なエントリを除去します。
     * @param controls 検証するコントロールレイアウト
     * @return 変更があった場合はtrue
     */
    fun sanitizeLayout(controls: CustomControls): Boolean {
        var madeChanges = sanitizeList(controls.mControlDataList!!)
        if (sanitizeList(controls.mDrawerDataList!!)) madeChanges = true
        if (sanitizeList(controls.mJoystickDataList!!)) madeChanges = true
        return madeChanges
    }
}
